subprojects/file-watching/src/main/java/org/gradle/internal/watch/registry/impl/WatchedHierarchies.java

```diff
@@ -0,0 +1,115 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.internal.watch.registry.impl;
+
+import com.google.common.annotations.VisibleForTesting;
+import org.gradle.internal.file.DefaultFileHierarchySet;
+import org.gradle.internal.file.FileHierarchySet;
+import org.gradle.internal.snapshot.CompleteFileSystemLocationSnapshot;
+import org.gradle.internal.snapshot.SnapshotHierarchy;
+
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.util.Comparator;
+import java.util.HashSet;
+import java.util.Set;
+import java.util.stream.Stream;
+
+public class WatchedHierarchies {
+    private Set<Path> watchedRoots = new HashSet<>();
+    private FileHierarchySet watchedHierarchies = DefaultFileHierarchySet.of();
+
+    public boolean contains(Path watchableHierarchy) {
+        return watchedHierarchies.contains(watchableHierarchy.toString());
+    }
+
+    public Set<Path> getWatchedRoots() {
+        return watchedRoots;
+    }
+
+    public void updateWatchedHierarchies(WatchableHierarchies watchableHierarchies, SnapshotHierarchy vfsRoot) {
+        watchedHierarchies = DefaultFileHierarchySet.of();
+        Stream<Path> hierarchiesWithSnapshots = watchableHierarchies.getWatchableHierarchies().stream()
+            .flatMap(watchableHierarchy -> {
+                if (watchedHierarchies.contains(watchableHierarchy.toString())) {
+                    return Stream.empty();
+                }
+                CheckIfNonEmptySnapshotVisitor checkIfNonEmptySnapshotVisitor = new CheckIfNonEmptySnapshotVisitor(watchableHierarchies);
+                vfsRoot.visitSnapshotRoots(watchableHierarchy.toString(), new FilterAlreadyCoveredSnapshotsVisitor(checkIfNonEmptySnapshotVisitor, watchedHierarchies));
+                if (checkIfNonEmptySnapshotVisitor.isEmpty()) {
+                    return Stream.empty();
+                }
+                watchedHierarchies = watchedHierarchies.plus(watchableHierarchy.toFile());
+                return checkIfNonEmptySnapshotVisitor.containsOnlyMissingFiles()
+                    ? Stream.of(locationOrFirstExistingAncestor(watchableHierarchy))
+                    : Stream.of(watchableHierarchy);
+            });
+
+        watchedRoots = resolveHierarchiesToWatch(hierarchiesWithSnapshots);
+    }
+
+    private Path locationOrFirstExistingAncestor(Path watchableHierarchy) {
+        if (Files.isDirectory(watchableHierarchy)) {
+            return watchableHierarchy;
+        }
+        return SnapshotWatchedDirectoryFinder.findFirstExistingAncestor(watchableHierarchy);
+    }
+
+    /**
+     * Filters out directories whose ancestor is also among the watched directories.
+     *
+     * The Stream of directories must not contain duplicates.
```

Comment:
This is almost like C++ documentation with their "undefined behaviors." :) Can we instead say what happens if there are duplicates? I also don't see why duplicates would cause problems.

Reply:
You are right, duplicates don't cause any problems. I removed the line from the Javadoc and added a test.